### PR TITLE
Feature/plugin 767 - Add BigQuery Plugin to load Array and Structs (And convert them to StructureRecords)

### PR DESCRIPTION
**Pull request by sechegaray**
Allow the BigQuery Plugin to load Array and Structs from BigQuery and convert them to StructureRecord to pass them downstream.

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java [hunk 1]
```diff
@@ -61,49 +62,97 @@ private BigQueryDataParser() {
...
       FieldValue fieldValue = fieldValues.get(fieldName);
-      FieldValue.Attribute attribute = fieldValue.getAttribute();
+      Attribute attribute = fieldValue.getAttribute();
       if (fieldValue.isNull()) {
         recordBuilder.set(fieldName, null);
         continue;
       }
 
-      if (attribute == FieldValue.Attribute.REPEATED) {
+      Schema.Field localSchemaField = schemaFieldList.get(i);
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java]: Are you assuming CDAP schema and BQ schema has same fields in same order?

**Review by tivv — COMMENTED** [review]
Please add a feature flag to this new capability. We should guard it with feature flag until after it's ready and tested. 
Also please give a good name to this PR

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java (hunk 1 above)
Comment (sechegaray) [reply src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java]: Refactored to avoid the assumption.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 2]
```diff
@@ -346,6 +346,11 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
         }
       }
 
+      if (!tableExists && Operation.UPSERT.equals(operation)) {
+        // If the table does not exist then an UPSERT operation becomes an INSERT operation
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: Is there a difference here?

Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: I get it now. Much cleaner.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java [hunk 3]
```diff
@@ -267,6 +268,68 @@ protected void validateInsertSchema(Table table, @Nullable Schema tableSchema, F
...
+    }
+
+    for (Schema.Field field: fields) {
+      String fieldName = prefix + field.getName();
+      if (depth == BigQueryTypeSize.Struct.MAX_DEPTH) {
+        collector.addFailure(
+          String.format("Field '%s' exceeds BigQuery maximum allowed depth of %d.",
+                        fieldName, BigQueryTypeSize.Struct.MAX_DEPTH),
+          "Please flatten the schema to contain fewer levels.");
+        continue;
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java:310]: Shouldn't this break/return here? If we discover the depth exceeds the threshold, we know this record is not processable and we should fail the pipeline.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java [hunk 4]
```diff
@@ -267,6 +268,68 @@ protected void validateInsertSchema(Table table, @Nullable Schema tableSchema, F
...
+          break;
+        case ARRAY:
+          if (fieldSchema.getComponentSchema() == null) {
+            break;
+          }
+          Schema componentSchema = BigQueryUtil.getNonNullableSchema(fieldSchema.getComponentSchema());
+          if (componentSchema.getType() == Schema.Type.RECORD) {
+            validateRecordDepth(componentSchema, collector, depth + 1, fieldName + ".");
+          } else {
+            validateRecordDepth(componentSchema, collector, depth, fieldName + ".");
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java:326]: So this means arrays are considered "flat" (same level) as a single value for a key?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkUtils.java [hunk 5]
```diff
@@ -272,16 +273,20 @@ private static BigQueryTableFieldSchema generateTableFieldSchema(Schema.Field fi
     fieldSchema.setType(type.name());
     if (type == LegacySQLTypeName.RECORD) {
       List<Schema.Field> schemaFields;
-      if (Schema.Type.ARRAY == field.getSchema().getType()) {
-        schemaFields = Objects.requireNonNull(field.getSchema().getComponentSchema()).getFields();
+      Schema fieldCdapSchema = BigQueryUtil.getNonNullableSchema(field.getSchema());
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkUtils.java:276]: Why not just `cdapSchema`?

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java [hunk 6]
```diff
@@ -63,49 +64,100 @@ private BigQueryDataParser() {
   }
 
   public static StructuredRecord getStructuredRecord(Schema schema, FieldList fields, FieldValueList fieldValues) {
+
     StructuredRecord.Builder recordBuilder = StructuredRecord.builder(schema);
-    for (Field field : fields) {
+    for (int i = 0; i < fields.size(); i++) {
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java]: Is this change required? It just makes accessing the list of fields O(N^2) instead of O(N)

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java [hunk 7]
```diff
@@ -63,49 +64,100 @@ private BigQueryDataParser() {
...
       FieldValue fieldValue = fieldValues.get(fieldName);
-      FieldValue.Attribute attribute = fieldValue.getAttribute();
+      Attribute attribute = fieldValue.getAttribute();
       if (fieldValue.isNull()) {
         recordBuilder.set(fieldName, null);
         continue;
       }
 
-      if (attribute == FieldValue.Attribute.REPEATED) {
+      Schema.Field localSchemaField = schema.getField(fieldName);
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java]: why not `localField` ?

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java [hunk 8]
```diff
@@ -63,49 +64,100 @@ private BigQueryDataParser() {
...
+      Schema localSchema;
+      if (localSchemaField != null) {
+        localSchema = localSchemaField.getSchema();
+      } else {
+        continue;
+      }
+      FieldList localFields = field.getSubFields();
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java]: `subFields` perhaps?

File: src/test/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtilTest.java [hunk 9]
```diff
@@ -95,9 +95,11 @@ public void testGetTableSchema() {
...
         Schema.nullableOf(Schema.of(Schema.Type.STRING)));
     Schema.Field cdapSubRecordField = Schema.Field.of("subStructure",
-        Schema.nullableOf(Schema.recordOf("structure.subStructure", cdapSubStringField)));
+        Schema.nullableOf(Schema.recordOf("structure.subStructure" +
+                                            Schema.recordOf(cdapSubStringField).getRecordName(), cdapSubStringField)));
     Schema.Field cdapRecordField = Schema.Field.of("structure",
-        Schema.nullableOf(Schema.recordOf("structure", cdapSubRecordField)));
+        Schema.nullableOf(Schema.recordOf("structure" +
+          Schema.recordOf(cdapSubRecordField).getRecordName(), cdapSubRecordField)));
     cdapFieldList.add(cdapRecordField);
```
Comment (fernst) [inline src/test/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtilTest.java:104]: For such a large change, I would like additional tests added to validate that this doesn't change existing behaviors.

**Review by fernst — COMMENTED** [review]
Some questions/comments

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java (hunk 3 above)
Comment (sechegaray) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java:310]: The pipeline will fail. I set continue so we can show the user ALL fields that are breaking this issue and not only the first one.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java (hunk 4 above)
Comment (sechegaray) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java:326]: Yes. Arrays in BQ are not a different level, they are the record but with mode "REPEATED."

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkUtils.java (hunk 5 above)
Comment (sechegaray) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkUtils.java:276]: The BQ field schema above is named `fieldSchema` so I wanted to connect both variables to make clear that they are referring to the same field.  Another solution to make it clearer would be to refactor the function and use `bqSchema` for the BQ field, and `cdapSchema` for cdap. But we haven't been using that standard across the code (prefixing bq)

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java (hunk 7 above)
Comment (sechegaray) [reply src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java]: Good point.

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java (hunk 8 above)
Comment (sechegaray) [reply src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryDataParser.java]: Good Idea.

File: src/test/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtilTest.java (hunk 9 above)
Comment (sechegaray) [reply src/test/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtilTest.java:104]: Added the test to check depth validation as discussed offline.

**Review by fernst — APPROVED** [review]
LGTM
